src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,666 @@
+                    match handle.response_code() {
+                        Ok(code) => Ok(CompletedDownload {
+                            response_code: code,
+                            data: download.data.take(),
+                            index_version: download
+                                .index_version
+                                .take()
+                                .unwrap_or_else(|| UNKNOWN.to_string()),
+                        }),
+                        Err(e) => Err(e),
+                    }
+                }
+                Err(e) => Err(e),
+            };
+            self.downloads.results.insert(download.path, result);
+        });
+        self.downloads.tick()?;
+
+        Ok(())
+    }
+
+    fn full_url(&self, path: &Path) -> String {
+        format!("{}/{}", self.url, path.display())
+    }
+
+    fn is_fresh(&self, path: &Path) -> bool {
+        if !self.requested_update {
+            trace!(
+                "using local {} as user did not request update",
+                path.display()
+            );
+            true
+        } else if self.config.cli_unstable().no_index_update {
+            trace!("using local {} in no_index_update mode", path.display());
+            true
+        } else if self.config.offline() {
+            trace!("using local {} in offline mode", path.display());
+            true
+        } else if self.fresh.contains(path) {
+            trace!("using local {} as it was already fetched", path.display());
+            true
+        } else {
+            debug!("checking freshness of {}", path.display());
+            false
+        }
+    }
+}
+
+impl<'cfg> RegistryData for HttpRegistry<'cfg> {
+    fn prepare(&self) -> CargoResult<()> {
+        Ok(())
+    }
+
+    fn index_path(&self) -> &Filesystem {
+        &self.index_path
+    }
+
+    fn assert_index_locked<'a>(&self, path: &'a Filesystem) -> &'a Path {
+        self.config.assert_package_cache_locked(path)
+    }
+
+    fn is_updated(&self) -> bool {
+        self.requested_update
+    }
+
+    fn load(
+        &mut self,
+        _root: &Path,
+        path: &Path,
+        index_version: Option<&str>,
+    ) -> Poll<CargoResult<LoadResponse>> {
+        trace!("load: {}", path.display());
+        if let Some(_token) = self.downloads.pending_ids.get(path) {
+            debug!("dependency is still pending: {}", path.display());
+            return Poll::Pending;
+        }
+
+        if let Some(index_version) = index_version {
+            trace!(
+                "local cache of {} is available at version `{}`",
+                path.display(),
+                index_version
+            );
+            if self.is_fresh(path) {
+                return Poll::Ready(Ok(LoadResponse::CacheValid));
+            }
+        } else if self.fresh.contains(path) {
+            debug!(
+                "cache did not contain previously downloaded file {}",
+                path.display()
+            );
+        }
```

Comment:
`self.fresh` only keeps track of which files have been downloaded in the current session, not the contents of the file.

The "cache did not contain previously downloaded file {}" is a case when the cache isn't working, so we go ahead and get the file again. This could possibly be made into an assert, since it shouldn't happen...